templates/default/components/Layout.jsx

```diff
@@ -1,13 +1,76 @@
 import Head from "next/head";
 import Link from "next/link";
+import { useCallback, useEffect, useState } from "react";
 
 import { siteConfig } from "../config/siteConfig";
 import { Nav } from "./Nav";
 
-export function Layout({ children }) {
+function useTableOfContents(tableOfContents) {
+  const [currentSection, setCurrentSection] = useState(tableOfContents[0]?.id);
+
+  const getHeadings = useCallback((toc) => {
+    return toc
+      .flatMap((node) => [node.id, ...node.children.map((child) => child.id)])
+      .map((id) => {
+        const el = document.getElementById(id);
+        if (!el) return;
+
+        const style = window.getComputedStyle(el);
+        const scrollMt = parseFloat(style.scrollMarginTop);
+
+        const top = window.scrollY + el.getBoundingClientRect().top - scrollMt;
+        return { id, top };
+      });
+  }, []);
```

Comment:
Does `[]` here means it will never be cached or it will always be cached? Does empty array do anything?

Reply:
Basically we are telling here to compute these values once since we just need them initially or else they will keep regenerating with the scroll top which makes the function expensive.